presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationPushDown.java

```diff
@@ -0,0 +1,222 @@
+
+        queryResult = queryRunner.executeWithQueryId(sessionWithUpperBoundPushDownEnabled, sql);
+        assertEquals(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputPositions(), 2);
+    }
+
+    @Test
+    public void testTimestampLogAppendModePushDown() throws ExecutionException, InterruptedException
+    {
+        Pair<String, String> timePair = createTimestampTestMessages(topicNameLogAppend);
+        DistributedQueryRunner queryRunner = (DistributedQueryRunner) getQueryRunner();
+        // ">= startTime" insure including index 2, "< endTime"  insure excluding index 4;
+        String sql = String.format("SELECT count(*) FROM default.%s WHERE _timestamp >= timestamp '%s' and _timestamp < timestamp '%s'",
+                topicNameLogAppend, timePair.first(), timePair.second());
+        ResultWithQueryId<MaterializedResult> queryResult = queryRunner.executeWithQueryId(getSession(), sql);
+        assertEquals(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputPositions(), 2);
+    }
+
+    private QueryInfo getQueryInfo(DistributedQueryRunner queryRunner, ResultWithQueryId<MaterializedResult> queryResult)
+    {
+        return queryRunner.getCoordinator().getQueryManager().getFullQueryInfo(queryResult.getQueryId());
+    }
+
+    private Pair<String, String> createTimestampTestMessages(String topicName) throws ExecutionException, InterruptedException
+    {
+        String startTime = null;
+        String endTime = null;
+        Future<RecordMetadata> lastSendFuture = Futures.immediateFuture(null);
+        try (KafkaProducer<Long, Object> producer = testingKafka.createProducer()) {
+            for (long messageNum = 0; messageNum < MESSAGE_NUM; messageNum++) {
+                long key = messageNum;
+                long value = messageNum;
+                lastSendFuture = producer.send(new ProducerRecord<>(topicName, key, value));
+                // Record timestamp to build expected timestamp
+                if (messageNum < TIMESTAMP_TEST_COUNT) {
+                    if (messageNum == TIMESTAMP_TEST_START_INDEX) {
+                        RecordMetadata r = lastSendFuture.get();
+                        startTime = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS")
+                                .format(LocalDateTime.ofInstant(Instant.ofEpochMilli(r.timestamp()), ZoneId.of("UTC")));
+                    }
+                    else if (messageNum == TIMESTAMP_TEST_END_INDEX) {
+                        RecordMetadata r = lastSendFuture.get();
+                        endTime = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS")
+                                .format(LocalDateTime.ofInstant(Instant.ofEpochMilli(r.timestamp()), ZoneId.of("UTC")));
+                    }
+                    else {
+                        lastSendFuture.get();
+                    }
+                    // Make timestamp more expected.
+                    Thread.sleep(20);
```

Comment:
```suggestion
                    // Sleep for a while to ensure different timestamps for different messages.
                    Thread.sleep(20);
```

Actually you can add an assertion that the timestamps for different messages are different in the tests which use `createTimestampTestMessages`